phoenix-core-server/src/main/java/org/apache/hadoop/hbase/regionserver/ScannerContextUtil.java

```diff
@@ -36,4 +36,28 @@ public static void incrementSizeProgress(ScannerContext sc, List<Cell> cells) {
                     cell.heapSize());
         }
     }
+
+    public static void updateMetrics(ScannerContext src, ScannerContext dst) {
+        if (src != null && dst != null && src.isTrackingMetrics() && dst.isTrackingMetrics()) {
+            for (Map.Entry<String, Long> entry : src.getMetrics().getMetricsMap().entrySet()) {
+                dst.metrics.addToCounter(entry.getKey(), entry.getValue());
+            }
+        }
+    }
+
+    public static ScannerContext copyNoLimitScanner(ScannerContext sc) {
+        return new ScannerContext(sc.keepProgress, null, sc.isTrackingMetrics());
+    }
+
+    public static void updateTimeProgress(ScannerContext sc) {
+        sc.updateTimeProgress();
+    }
+
+    /**
+     * Increment the heapSize progress to a large value for server paging.
+     * Not using Long.MAX_VALUE since that can cause overflow.
+     */
+    public static void incrementHeapProgressForPaging(ScannerContext sc) {
+        sc.incrementSizeProgress(0, Long.MAX_VALUE/2);
+    }
```

Comment:
@kadirozde @palashc This is nice trick. Is `Long.MAX_VALUE/2` guaranteed to cause large enough size for HBase to return response?

Reply:
Yes, 4096 PB will be beyond any practical heap size for a very long time.

Reply:
However, you have a good point. @palashc, the default heap size is -1, which means it is not enforced. What happens in that case?

Reply:
@palashc, we may have to set the heap limit to Long.MAX_VALUE/2 on the scanner context on the way down and increment heap progress by Long.MAX_VALUE/2 when dummy result is received. Also we need to make sure that this manipulation of the scanner context is done only once for a given scanner.

Reply:
@palashc it seems we have better alternative i.e. instead of size limit, we can enforce time limit reached by using

```
  // Set this to true will have the same behavior with reaching the time limit.
  // This is used when you want to make the current RSRpcService.scan returns immediately. For
  // example, when we want to switch from pread to stream, we can only do it after the rpc call is
  // returned.
  private boolean returnImmediately;
```

```
  void returnImmediately() {
    this.returnImmediately = true;
  }
```

This enforces `timeLimitReached`. HBase uses three factors to determine whether to return immediately:
1. sizeLimitReached
2. timeLimitReached
3. resultsLimitReached

Reply:
@palashc, the above by @virajjasani is a much simpler and better approach. Please verify that it works.

Reply:
Made the change in the latest commit after verifying using the new IT I added in ServerPagingIT. 
Thanks @virajjasani , great find!

Reply:
@kadirozde @virajjasani Please let me know if you would be able to review the rest of the changes in this PR regarding scannerContext.